File: Algorithms/3sumClosest.java
import java.util.*;
public class Solution {
    public int threeSumClosest(int[] num, int target) {
        Arrays.sort(num);
        int n = num.length;
        int lowestGap = Integer.MAX_VALUE;
        int closestSum = 0;
        for (int i = 0; i < n; i++) {
            while (i > 0 && i < n && num[i] == num[i-1])
                i++;
            int j = i + 1;
            int k = n - 1;
            while (j < k) {
                int total = num[i] + num[j] + num[k];
                int gap = Math.abs(target - total);
                if (gap < lowestGap) {
                    lowestGap = gap;
                    closestSum = total;
                }
                if (total < target) {
                    j++;
                    while (j < k && num[j] == num[j-1])
                        j++;
                } else if (total > target) {
                    k--;
                    while (k > j && num[k] == num[k+1])
                        k--;
                } else
                    return closestSum;
            }
        }
        return closestSum;
    }
}

File: Algorithms/3sumClosest.py
class Solution:
    # @return an integer
    def threeSumClosest(self, num, target):
        result = set()
        num = sorted(num)
        n = len(num)
        i = 0
        closestSum = num[0] + num[1] + num[2]
        min_gap = abs(closestSum - target)
        while i < n:
            while i > 0 and i < n and num[i] == num[i-1]:
                i += 1
            j = i + 1
            k = n - 1
            while j < k:
                total = num[i] + num[j] + num[k]
                gap = abs(total - target)
                if gap < min_gap:
                    closestSum = total
                    min_gap = gap
                if total < target:
                    j += 1
                    while j < k and num[j] == num[j-1]:
                        j += 1
                elif total > target:
                    k -= 1
                    while k > j and num[k] == num[k+1]:
                        k -= 1
                else:
                    return closestSum
            i += 1
        return closestSum
